File: jscomp/others/belt_MutableMapInt.ml
# 2 "others/mapm.cppo.ml"
module I = Belt_internalMapInt
type key = int
# 10 "others/mapm.cppo.ml"
module N = Belt_internalAVLtree
module A = Belt_Array 




type 'a t = {
  mutable data : 'a I.t
} 


let make () = {data = None}
let isEmpty m = N.isEmpty m.data
let clear m =  m.data<- None
(* let singleton k v = t ~data:(N.singleton k v) *)

let minKeyUndefined m = N.minKeyUndefined m.data
let minKey m = N.minKey m.data
let maxKeyUndefined m = N.maxKeyUndefined m.data
let maxKey m = N.maxKey m.data
let minimum m = N.minimum m.data
let minUndefined m = N.minUndefined m.data
let maximum m = N.maximum m.data
let maxUndefined m = N.maxUndefined m.data

let set (m : _ t) k v = 
  let old_data = m.data in 
  let v = I.addMutate old_data k v in 
  if v != old_data then 
    m.data <-v 

let forEachU d f = N.forEachU d.data f
let forEach d f = forEachU d (fun[@bs] a b -> f a b)    
let mapU d f = {data = N.mapU d.data f}
let map d f = mapU d (fun[@bs] a -> f a )    
let mapWithKeyU d f = { data = (N.mapWithKeyU d.data f)}
let mapWithKey d f = mapWithKeyU d (fun [@bs] a b -> f a b)    
let reduceU d acc f  = N.reduceU d.data acc f
let reduce d acc f = reduceU d acc (fun[@bs] a b c -> f a b c)    
let everyU d f = N.everyU d.data f
let every d f = everyU d (fun[@bs] a b -> f a b)    
let someU d f = N.someU d.data f    
let some d f = someU d (fun[@bs] a b -> f a b)
let size d = N.size d.data
let toList d = N.toList d.data
let toArray d = N.toArray d.data
let keysToArray d = N.keysToArray d.data
let valuesToArray d = N.valuesToArray d.data
let checkInvariantInternal d = N.checkInvariantInternal d.data
let has d v = I.has d.data v 


let rec removeMutateAux nt (x : key)= 
  let k = nt.N.key in 
  if x = k then 
    let {N.left = l; right = r} = nt in       
    match l, r with
    | None, _ -> r  
    | _, None -> l 
    | _,  Some nr ->  
      nt.right <- (N.removeMinAuxWithRootMutate nt nr);
      Some (N.balMutate nt)
  else 
    begin 
      if x < k then 
        match  nt.left with         
        | None -> Some nt 
        | Some l ->
          nt.left <- (removeMutateAux l x );
          Some (N.balMutate nt)
      else 
        match  nt.right with 
        | None -> Some nt 
        | Some r -> 
          nt.right <- (removeMutateAux r x);
          Some (N.balMutate nt)
    end

let remove d v = 
  let oldRoot = d.data in 
  match  oldRoot with 
  | None -> ()
  | Some root -> 
    let newRoot = removeMutateAux root v in 
    if newRoot != oldRoot then 
      d.data <- newRoot   


let rec updateDone t (x : key)  f  =   
  match  t with 
  | None ->
    (match f None [@bs] with
    | Some data -> N.singleton x data
    | None -> t)
  | Some nt -> 
    let k = nt.N.key in 
    (* let  c = (Belt_Cmp.getCmpInternal cmp) x k [@bs] in   *)
    if k = x then begin     
      match f (Some nt.value) [@bs] with
      | None ->
        let {N.left = l; right = r} = nt in
        begin match  l,  r with
          | None,  _ -> r
          | _, None  -> l
          | _, Some nr ->
            nt.right <- (N.removeMinAuxWithRootMutate nt nr);
            Some (N.balMutate nt)
        end
      | Some data -> 
        nt.value <- data;
        Some nt
    end      
    else
      let {N.left = l; right = r} = nt in 
      (if x < k then                   
         let ll = updateDone  l x f in
         nt.left <- ll
       else   
         nt.right <- (updateDone  r x f);
      );
      Some (N.balMutate nt)
        
let updateU t x f =       
  let oldRoot = t.data in 
  let newRoot = updateDone oldRoot x f  in 
  if newRoot != oldRoot then 
    t.data <- newRoot 
let update t x f = updateU t x (fun[@bs] a -> f a )
let rec removeArrayMutateAux t xs i len   =  
  if i < len then 
    let ele = A.getUnsafe xs i in 
    let u = removeMutateAux t ele  in 
    match  u with 
    | None -> None
    | Some t -> removeArrayMutateAux t xs (i+1) len 
  else Some t    

let removeMany (d : _ t) xs =  
  let oldRoot = d.data in 
  match  oldRoot with 
  | None -> ()
  | Some nt -> 
    let len = A.length xs in 
    let newRoot = removeArrayMutateAux nt xs 0 len in 
    if newRoot != oldRoot then 
      d.data <- newRoot



(* let split = I.split  *)
(* let merge = I.merge  *)


let fromArray xs = 
  { data = I.fromArray xs }

let cmpU d0 d1 f = 
  I.cmpU d0.data d1.data f
let cmp d0 d1 f = cmpU d0 d1 (fun[@bs] a b -> f a b)          

let eqU d0 d1 f = 
  I.eqU d0.data d1.data f
let eq d0 d1 f = eqU d0 d1 (fun[@bs] a b -> f a b)    

let get d x =   I.get d.data x 
let getUndefined d x = I.getUndefined d.data x 
let getWithDefault d x def = I.getWithDefault d.data x def  
let getExn d x = I.getExn d.data x 

File: lib/es6/Belt_MutableMapInt.js


import * as Primitive_option from "./Primitive_option.js";
import * as Belt_internalMapInt from "./Belt_internalMapInt.js";
import * as Belt_internalAVLtree from "./Belt_internalAVLtree.js";

function make() {
  return {
    data: undefined
  };
}

function isEmpty(m) {
  let x = m.data;
  return x === undefined;
}

function clear(m) {
  m.data = undefined;
}

function minKeyUndefined(m) {
  return Belt_internalAVLtree.minKeyUndefined(m.data);
}

function minKey(m) {
  return Belt_internalAVLtree.minKey(m.data);
}

function maxKeyUndefined(m) {
  return Belt_internalAVLtree.maxKeyUndefined(m.data);
}

function maxKey(m) {
  return Belt_internalAVLtree.maxKey(m.data);
}

function minimum(m) {
  return Belt_internalAVLtree.minimum(m.data);
}

function minUndefined(m) {
  return Belt_internalAVLtree.minUndefined(m.data);
}

function maximum(m) {
  return Belt_internalAVLtree.maximum(m.data);
}

function maxUndefined(m) {
  return Belt_internalAVLtree.maxUndefined(m.data);
}

function set(m, k, v) {
  let old_data = m.data;
  let v$1 = Belt_internalMapInt.addMutate(old_data, k, v);
  if (v$1 !== old_data) {
    m.data = v$1;
    return;
  }
  
}

function forEach(d, f) {
  Belt_internalAVLtree.forEach(d.data, f);
}

function map(d, f) {
  return {
    data: Belt_internalAVLtree.map(d.data, f)
  };
}

function mapWithKey(d, f) {
  return {
    data: Belt_internalAVLtree.mapWithKey(d.data, f)
  };
}

function reduce(d, acc, f) {
  return Belt_internalAVLtree.reduce(d.data, acc, f);
}

function every(d, f) {
  return Belt_internalAVLtree.every(d.data, f);
}

function some(d, f) {
  return Belt_internalAVLtree.some(d.data, f);
}

function size(d) {
  return Belt_internalAVLtree.size(d.data);
}

function toList(d) {
  return Belt_internalAVLtree.toList(d.data);
}

function toArray(d) {
  return Belt_internalAVLtree.toArray(d.data);
}

function keysToArray(d) {
  return Belt_internalAVLtree.keysToArray(d.data);
}

function valuesToArray(d) {
  return Belt_internalAVLtree.valuesToArray(d.data);
}

function checkInvariantInternal(d) {
  Belt_internalAVLtree.checkInvariantInternal(d.data);
}

function has(d, v) {
  return Belt_internalMapInt.has(d.data, v);
}

function removeMutateAux(nt, x) {
  let k = nt.k;
  if (x === k) {
    let l = nt.l;
    let r = nt.r;
    if (l !== undefined) {
      if (r !== undefined) {
        nt.r = Belt_internalAVLtree.removeMinAuxWithRootMutate(nt, r);
        return Belt_internalAVLtree.balMutate(nt);
      } else {
        return l;
      }
    } else {
      return r;
    }
  }
  if (x < k) {
    let l$1 = nt.l;
    if (l$1 !== undefined) {
      nt.l = removeMutateAux(l$1, x);
      return Belt_internalAVLtree.balMutate(nt);
    } else {
      return nt;
    }
  }
  let r$1 = nt.r;
  if (r$1 !== undefined) {
    nt.r = removeMutateAux(r$1, x);
    return Belt_internalAVLtree.balMutate(nt);
  } else {
    return nt;
  }
}

function remove(d, v) {
  let oldRoot = d.data;
  if (oldRoot === undefined) {
    return;
  }
  let newRoot = removeMutateAux(oldRoot, v);
  if (newRoot !== oldRoot) {
    d.data = newRoot;
    return;
  }
  
}

function updateDone(t, x, f) {
  if (t !== undefined) {
    let k = t.k;
    if (k === x) {
      let data = f(Primitive_option.some(t.v));
      if (data !== undefined) {
        t.v = Primitive_option.valFromOption(data);
        return t;
      }
      let l = t.l;
      let r = t.r;
      if (l !== undefined) {
        if (r !== undefined) {
          t.r = Belt_internalAVLtree.removeMinAuxWithRootMutate(t, r);
          return Belt_internalAVLtree.balMutate(t);
        } else {
          return l;
        }
      } else {
        return r;
      }
    }
    let l$1 = t.l;
    let r$1 = t.r;
    if (x < k) {
      let ll = updateDone(l$1, x, f);
      t.l = ll;
    } else {
      t.r = updateDone(r$1, x, f);
    }
    return Belt_internalAVLtree.balMutate(t);
  }
  let data$1 = f(undefined);
  if (data$1 !== undefined) {
    return Belt_internalAVLtree.singleton(x, Primitive_option.valFromOption(data$1));
  } else {
    return t;
  }
}

function update(t, x, f) {
  let oldRoot = t.data;
  let newRoot = updateDone(oldRoot, x, f);
  if (newRoot !== oldRoot) {
    t.data = newRoot;
    return;
  }
  
}

function removeArrayMutateAux(_t, xs, _i, len) {
  while (true) {
    let i = _i;
    let t = _t;
    if (i >= len) {
      return t;
    }
    let ele = xs[i];
    let u = removeMutateAux(t, ele);
    if (u === undefined) {
      return;
    }
    _i = i + 1 | 0;
    _t = u;
    continue;
  };
}

function removeMany(d, xs) {
  let oldRoot = d.data;
  if (oldRoot === undefined) {
    return;
  }
  let len = xs.length;
  let newRoot = removeArrayMutateAux(oldRoot, xs, 0, len);
  if (newRoot !== oldRoot) {
    d.data = newRoot;
    return;
  }
  
}

function fromArray(xs) {
  return {
    data: Belt_internalMapInt.fromArray(xs)
  };
}

function cmp(d0, d1, f) {
  return Belt_internalMapInt.cmp(d0.data, d1.data, f);
}

function eq(d0, d1, f) {
  return Belt_internalMapInt.eq(d0.data, d1.data, f);
}

function get(d, x) {
  return Belt_internalMapInt.get(d.data, x);
}

function getUndefined(d, x) {
  return Belt_internalMapInt.getUndefined(d.data, x);
}

function getWithDefault(d, x, def) {
  return Belt_internalMapInt.getWithDefault(d.data, x, def);
}

function getExn(d, x) {
  return Belt_internalMapInt.getExn(d.data, x);
}

let cmpU = cmp;

let eqU = eq;

let forEachU = forEach;

let reduceU = reduce;

let everyU = every;

let someU = some;

let updateU = update;

let mapU = map;

let mapWithKeyU = mapWithKey;

export {
  make,
  clear,
  isEmpty,
  has,
  cmpU,
  cmp,
  eqU,
  eq,
  forEachU,
  forEach,
  reduceU,
  reduce,
  everyU,
  every,
  someU,
  some,
  size,
  toList,
  toArray,
  fromArray,
  keysToArray,
  valuesToArray,
  minKey,
  minKeyUndefined,
  maxKey,
  maxKeyUndefined,
  minimum,
  minUndefined,
  maximum,
  maxUndefined,
  get,
  getUndefined,
  getWithDefault,
  getExn,
  checkInvariantInternal,
  remove,
  removeMany,
  set,
  updateU,
  update,
  mapU,
  map,
  mapWithKeyU,
  mapWithKey,
}
/* No side effect */
